src/ddspdrum/module.py

```diff
@@ -145,24 +226,29 @@ def _ramp(self, duration: float):
         t = np.linspace(
             0, duration, int(round(duration * self.control_rate)), endpoint=False
         )
-        return (t / duration) ** self.alpha
+        return (t / duration) ** self.parameters['alpha'].value
 
     @property
     def attack(self):
-        return self._ramp(self.a)
+        return self._ramp(self.parameters['attack'].value)
 
     @property
     def decay(self):
         # `d`-length reverse ramp, scaled and shifted to descend from 1 to `s`.
-        return self._ramp(self.d)[::-1] * (1 - self.s) + self.s
+        decay = self.parameters['decay'].value
+        sustain = self.parameters['sustain'].value
+        return self._ramp(decay)[::-1] * (1 - sustain) + sustain
 
     def sustain(self, duration):
-        return np.full(round(int(duration * CONTROL_RATE)), fill_value=self.s)
+        return np.full(round(int(duration * CONTROL_RATE)),
+                       fill_value=self.parameters['sustain'].value)
 
     @property
     def release(self):
         # `r`-length reverse ramp, scaled and shifted to descend from `s` to 0.
-        return self._ramp(self.r)[::-1] * self.s
+        sustain = self.parameters['sustain'].value
+        release = self.parameters['release'].value
+        return self._ramp(release)[::-1] * sustain
```

Comment:
One downside is these parameters take up more room ... maybe there is a way around this?

Reply:
We could use:
```
def p(self, name): return self.parameters[name].value
```
which I think is an okay tradeoff given how common this idiom is in our code. Then you just do `self.p('sustain')`.

If we want to get rid of the string quotes, well we could use python black magic fuckery.

This won't work: at the end of __init__ there could be a generic SynthModule method that is called that does:

```
for p in self.parameters:
    self.__dict__[p] = self.parameters[p].value
```

the reason that won't work is because it does assign-by-value not assign-by-reference.

Instead, there might be a way to create @get and @set properties using lambda for every parameter. This might be too tricky, and perhaps just a `self.p('name')` is best

Reply:
`def p(self, name): return self.parameters[name].value` -- I like this

Reply:
yeah easy and concise

Reply:
+1

Reply:
Let's just figure out whether we want to use tuples or connect_parameter and strings for compound parameter addressing, before we tackle this.